cmd/topicmappr/commands/scale.go

```diff
@@ -0,0 +1,219 @@
+}
+
+func scale(cmd *cobra.Command, _ []string) {
+	bootstrap(cmd)
+
+	// ZooKeeper init.
+	zk, err := initZooKeeper(cmd)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	defer zk.Close()
+
+	// Get broker and partition metadata.
+	checkMetaAge(cmd, zk)
+	brokerMeta := getBrokerMeta(cmd, zk, true)
+	partitionMeta := getPartitionMeta(cmd, zk)
+
+	// Get the current partition map.
+	partitionMapIn, err := kafkazk.PartitionMapFromZK(Config.topics, zk)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// Exclude any topics that are pending deletion.
+	pending := stripPendingDeletes(partitionMapIn, zk)
+
+	// Exclude any explicit exclusions.
+	excluded := removeTopics(partitionMapIn, Config.topicsExclude)
+
+	// Print topics matched to input params.
+	printTopics(partitionMapIn)
+
+	// Print if any topics were excluded due to pending deletion.
+	printExcludedTopics(pending, excluded)
+
+	// Get a broker map.
+	brokersIn := kafkazk.BrokerMapFromPartitionMap(partitionMapIn, brokerMeta, false)
+
+	// Validate all broker params, get a copy of the
+	// broker IDs targeted for partition offloading.
+	offloadTargets := validateBrokersForScale(cmd, brokersIn, brokerMeta)
+
+	// Sort offloadTargets by storage free ascending.
+	sort.Sort(offloadTargetsBySize{t: offloadTargets, bm: brokersIn})
+
+	partitionLimit, _ := cmd.Flags().GetInt("partition-limit")
+	partitionSizeThreshold, _ := cmd.Flags().GetInt("partition-size-threshold")
+	tolerance, _ := cmd.Flags().GetFloat64("tolerance")
+	localityScoped, _ := cmd.Flags().GetBool("locality-scoped")
+	verbose, _ := cmd.Flags().GetBool("verbose")
+
+	// Generate reassignmentBundles for a scale up.
+	params := computeReassignmentBundlesParams{
+		offloadTargets:         offloadTargets,
+		tolerance:              tolerance,
+		partitionMap:           partitionMapIn,
+		partitionMeta:          partitionMeta,
+		brokerMap:              brokersIn,
+		partitionLimit:         partitionLimit,
+		partitionSizeThreshold: partitionSizeThreshold,
+		localityScoped:         localityScoped,
+		verbose:                verbose,
+	}
+
+	results := computeReassignmentBundles(params)
+
+	// Merge all results into a slice.
+	resultsByRange := []reassignmentBundle{}
+	for r := range results {
+		resultsByRange = append(resultsByRange, r)
+	}
+
+	// Sort the scale results by range ascending.
+	sort.Slice(resultsByRange, func(i, j int) bool {
+		switch {
+		case resultsByRange[i].storageRange < resultsByRange[j].storageRange:
+			return true
+		case resultsByRange[i].storageRange > resultsByRange[j].storageRange:
+			return false
+		}
+
+		return resultsByRange[i].stdDev < resultsByRange[j].stdDev
+	})
+
+	// Chose the results with the lowest range.
+	m := resultsByRange[0]
+	partitionMapOut, brokersOut, relos := m.partitionMap, m.brokers, m.relocations
+
+	// Print parameters used for scale decisions.
+	printReassignmentParams(cmd, resultsByRange, brokersIn, m.tolerance)
+
+	// Optimize leaders.
```

Comment:
is this expected to have in a scale operation?

Reply:
It's optional, so just keeping it consistent with the other commands (in that you can decide to also perform a leadership optimization whether a scale, rebalance, or rebuild)